examples/src/read-data.c

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2020, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ *
+ * Copyright 2019, UT-Battelle, LLC.
+ *
+ * LLNL-CODE-741539
+ * All rights reserved.
+ *
+ * This is the license for UnifyFS.
+ * For details, see https://github.com/LLNL/UnifyFS.
+ * Please read https://github.com/LLNL/UnifyFS/LICENSE for full license text.
+ */
+/* read-data: This program aims to test reading data from a file with arbitrary
+ * offset and length. This program can either interactively (only specifying
+ * filename) or non-interactively (specifying filename with offset and length).
+ */
+#include <config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <errno.h>
+#include <limits.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <libgen.h>
+#include <getopt.h>
+#include <unifyfs.h>
+
+#include "testutil.h"
+
+#define DEFAULT_MOUNTPOINT "/unifyfs"
+
+static char filename[PATH_MAX];
+static char mountpoint[PATH_MAX];
+
+static char* buf;
+static unsigned int bufsize;
+
+static int parse_line(char* line, unsigned long* offset, unsigned long* length)
+{
+    char* pos = NULL;
+
+    line[strlen(line)-1] = '\0';
+
+    if (strncmp("quit", line, strlen("quit")) == 0) {
+        return 1;
+    }
+
+    pos = strchr(line, ',');
+    if (!pos) {
+        return -1;
+    }
+
+    *pos = '\0';
+    pos++;
+
+    *offset = strtoul(line, NULL, 0);
+    *length = strtoul(pos, NULL, 0);
+
+    return 0;
+}
+
+static void alloc_buf(unsigned long length)
+{
+    if (!buf) {
+        bufsize = length;
+        buf = malloc(bufsize);
+    } else {
+        if (bufsize < length) {
+            buf = realloc(buf, length);
+        }
+    }
+
+    if (!buf) {
+        perror("failed to allocate buffer");
+        exit(1);
+    }
+}
+
+static void do_pread(int fd, unsigned long length, unsigned long offset)
```

Comment:
why aren't you using the actual types for `pread()` parameters (i.e., `off_t offset` and `size_t length`)?

Reply:
Thanks, @MichaelBrim. I've updated it.